File: services/api/src/test/java/com/voltpilot/api/uems/UemsSchemaLaeufer.java
package com.voltpilot.api.uems;

import com.fasterxml.jackson.databind.JsonNode;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.Pattern;

/**
 * Ein kleiner Läufer über die Teilmenge von JSON-Schema draft 2020-12, die die
 * UEMS-Vektor-Schemas benutzen ({@code uems-referenzunternehmen.schema.json},
 * {@code messwert-herkunft.schema.json}): {@code type}, {@code required},
 * {@code properties}, {@code additionalProperties:false}, {@code items},
 * {@code enum}, {@code const}, {@code pattern}, {@code minItems},
 * {@code minLength}, {@code maxLength}, {@code minimum}, {@code maximum} und
 * {@code $ref} auf {@code #/$defs/…}. Das Projekt hat keine Schema-Bibliothek
 * (siehe {@code services/api/pom.xml}); der Läufer ersetzt keine — er hält eine
 * Datei an genau den Regeln fest, die ihr Schema aufschreibt, und sagt bei jedem
 * Verstoß den Pfad. Ein Schlüsselwort außerhalb dieser Teilmenge (etwa
 * {@code maxItems}) prüft er NICHT; wer sich darauf verlässt, prüft es im Test
 * selbst.
 *
 * <p>Der TS-Zwilling des Referenzunternehmens
 * ({@code frontend/portal/src/uemsReferenzunternehmen.test.ts}) trägt denselben
 * Läufer.
 */
final class UemsSchemaLaeufer {
    private final JsonNode wurzel;
    private final List<String> fehler;
    private final Map<String, Pattern> muster = new HashMap<>();

    UemsSchemaLaeufer(JsonNode wurzel, List<String> fehler) {
        this.wurzel = wurzel;
        this.fehler = fehler;
    }

    /** Alle Verstöße von {@code daten} gegen {@code schema}, je mit Pfad; leer heißt: hält. */
    static List<String> verstoesse(JsonNode daten, JsonNode schema) {
        List<String> fehler = new ArrayList<>();
        new UemsSchemaLaeufer(schema, fehler).pruefe(daten, schema, "$");
        return fehler;
    }

    void pruefe(JsonNode wert, JsonNode schema, String pfad) {
        if (schema.has("$ref")) {
            String ref = schema.get("$ref").asText();
            JsonNode ziel = wurzel.at(ref.substring(1));
            if (ziel.isMissingNode()) {
                fehler.add(pfad + ": unbekannter Schema-Verweis " + ref);
                return;
            }
            pruefe(wert, ziel, pfad);
            return;
        }
        if (schema.has("type") && !typPasst(wert, schema.get("type"))) {
            fehler.add(pfad + ": Typ " + typVon(wert) + " passt nicht zu " + schema.get("type"));
            return;
        }
        if (schema.has("const") && !wert.equals(schema.get("const"))) {
            fehler.add(pfad + ": " + wert + " ist nicht " + schema.get("const"));
        }
        if (schema.has("enum") && !wert.isNull()) {
            boolean drin = false;
            for (JsonNode e : schema.get("enum")) {
                drin |= e.equals(wert);
            }
            if (!drin) {
                fehler.add(pfad + ": " + wert + " steht nicht im Vokabular " + schema.get("enum"));
            }
        }
        if (wert.isTextual()) {
            pruefeText(wert.asText(), schema, pfad);
        }
        if (wert.isNumber()) {
            if (schema.has("minimum") && wert.asDouble() < schema.get("minimum").asDouble()) {
                fehler.add(pfad + ": " + wert + " unter dem Mindestwert");
            }
            if (schema.has("maximum") && wert.asDouble() > schema.get("maximum").asDouble()) {
                fehler.add(pfad + ": " + wert + " über dem Höchstwert");
            }
        }
        if (wert.isArray()) {
            if (schema.has("minItems") && wert.size() < schema.get("minItems").asInt()) {
                fehler.add(pfad + ": zu wenige Einträge");
            }
            if (schema.has("items")) {
                for (int i = 0; i < wert.size(); i++) {
                    pruefe(wert.get(i), schema.get("items"), pfad + "[" + i + "]");
                }
            }
        }
        if (wert.isObject()) {
            pruefeObjekt(wert, schema, pfad);
        }
    }

    private void pruefeText(String s, JsonNode schema, String pfad) {
        if (schema.has("pattern")
                && !muster.computeIfAbsent(schema.get("pattern").asText(), Pattern::compile)
                        .matcher(s).find()) {
            fehler.add(pfad + ": „" + s + "“ passt nicht zum Muster "
                    + schema.get("pattern").asText());
        }
        if (schema.has("minLength") && s.length() < schema.get("minLength").asInt()) {
            fehler.add(pfad + ": zu kurz");
        }
        if (schema.has("maxLength") && s.length() > schema.get("maxLength").asInt()) {
            fehler.add(pfad + ": zu lang");
        }
    }

    private void pruefeObjekt(JsonNode wert, JsonNode schema, String pfad) {
        for (JsonNode p : schema.path("required")) {
            if (!wert.has(p.asText())) {
                fehler.add(pfad + ": Pflichtfeld " + p.asText() + " fehlt");
            }
        }
        JsonNode props = schema.path("properties");
        JsonNode zusatz = schema.path("additionalProperties");
        Set<String> bekannt = new HashSet<>();
        props.fieldNames().forEachRemaining(bekannt::add);
        wert.fields().forEachRemaining(e -> {
            if (bekannt.contains(e.getKey())) {
                pruefe(e.getValue(), props.get(e.getKey()), pfad + "." + e.getKey());
            } else if (zusatz.isObject()) {
                pruefe(e.getValue(), zusatz, pfad + "." + e.getKey());
            } else if (zusatz.isBoolean() && !zusatz.asBoolean()) {
                fehler.add(pfad + ": unbekanntes Feld " + e.getKey());
            }
        });
    }

    private static boolean typPasst(JsonNode wert, JsonNode typ) {
        if (typ.isArray()) {
            for (JsonNode t : typ) {
                if (einTypPasst(wert, t.asText())) {
                    return true;
                }
            }
            return false;
        }
        return einTypPasst(wert, typ.asText());
    }

    private static boolean einTypPasst(JsonNode wert, String typ) {
        return switch (typ) {
            case "object" -> wert.isObject();
            case "array" -> wert.isArray();
            case "string" -> wert.isTextual();
            case "integer" -> wert.isIntegralNumber();
            case "number" -> wert.isNumber();
            case "boolean" -> wert.isBoolean();
            case "null" -> wert.isNull();
            default -> false;
        };
    }

    private static String typVon(JsonNode wert) {
        if (wert.isObject()) {
            return "object";
        }
        if (wert.isArray()) {
            return "array";
        }
        if (wert.isTextual()) {
            return "string";
        }
        if (wert.isIntegralNumber()) {
            return "integer";
        }
        if (wert.isNumber()) {
            return "number";
        }
        if (wert.isBoolean()) {
            return "boolean";
        }
        return "null";
    }
}

File: frontend/portal/src/test/uemsSchemaLaeufer.ts
/**
 * Ein kleiner Läufer über die Teilmenge von JSON-Schema draft 2020-12, die die
 * UEMS-Vektor-Schemas unter `docs/contracts/v2/` benutzen: `type`, `required`,
 * `properties`, `additionalProperties` (false oder Schema), `items`, `enum`,
 * `const`, `pattern`, `minItems`, `minLength`, `maxLength`, `minimum`,
 * `maximum`, `anyOf` und `$ref` auf `#/$defs/…`. Das Projekt hat keine
 * Schema-Bibliothek; der Läufer ersetzt keine — er hält eine Datei an genau den
 * Regeln fest, die ihr Schema aufschreibt, und sagt bei jedem Verstoß den Pfad.
 *
 * Der Java-Zwilling ist `services/api .../uems/UemsSchemaLaeufer` (Testquelle).
 */

type Json = any;

const typVon = (v: Json): string => {
  if (v === null) return 'null';
  if (Array.isArray(v)) return 'array';
  if (typeof v === 'number') return Number.isInteger(v) ? 'integer' : 'number';
  return typeof v === 'object' ? 'object' : typeof v;
};

const passt = (v: Json, t: string): boolean => (t === 'number' ? typeof v === 'number' : typVon(v) === t);

/**
 * Alle Verstöße von `daten` gegen `teilschema` (Vorgabe: die Wurzel), je mit
 * Pfad; `$ref` wird gegen `wurzel` aufgelöst. Leer heißt: hält.
 */
export function schemaVerstoesse(daten: Json, wurzel: Json, teilschema: Json = wurzel): string[] {
  const pruefe = (wert: Json, s: Json, pfad: string): string[] => {
    if (s.$ref) {
      const ziel = (s.$ref as string)
        .slice(2)
        .split('/')
        .reduce((o: Json, t: string) => (o == null ? o : o[t.replace(/~1/g, '/').replace(/~0/g, '~')]), wurzel);
      if (!ziel) return [`${pfad}: unbekannter Schema-Verweis ${s.$ref}`];
      return pruefe(wert, ziel, pfad);
    }
    const fehler: string[] = [];
    if (s.anyOf && !(s.anyOf as Json[]).some((zweig) => pruefe(wert, zweig, pfad).length === 0)) {
      fehler.push(`${pfad}: ${JSON.stringify(wert)} passt zu keinem Zweig von anyOf`);
    }
    if (s.type) {
      const typen: string[] = Array.isArray(s.type) ? s.type : [s.type];
      if (!typen.some((t) => passt(wert, t))) return [...fehler, `${pfad}: Typ ${typVon(wert)} passt nicht zu ${typen.join('|')}`];
    }
    if ('const' in s && JSON.stringify(wert) !== JSON.stringify(s.const)) {
      fehler.push(`${pfad}: ${JSON.stringify(wert)} ist nicht ${JSON.stringify(s.const)}`);
    }
    if (s.enum && wert !== null && !s.enum.includes(wert)) {
      fehler.push(`${pfad}: ${JSON.stringify(wert)} steht nicht im Vokabular`);
    }
    if (typeof wert === 'string') {
      if (s.pattern && !new RegExp(s.pattern, 'u').test(wert)) {
        fehler.push(`${pfad}: „${wert}“ passt nicht zum Muster ${s.pattern}`);
      }
      if (s.minLength != null && wert.length < s.minLength) fehler.push(`${pfad}: zu kurz`);
      if (s.maxLength != null && wert.length > s.maxLength) fehler.push(`${pfad}: zu lang`);
    }
    if (typeof wert === 'number') {
      if (s.minimum != null && wert < s.minimum) fehler.push(`${pfad}: unter dem Mindestwert`);
      if (s.maximum != null && wert > s.maximum) fehler.push(`${pfad}: über dem Höchstwert`);
    }
    if (Array.isArray(wert)) {
      if (s.minItems != null && wert.length < s.minItems) fehler.push(`${pfad}: zu wenige Einträge`);
      if (s.items) wert.forEach((v, i) => fehler.push(...pruefe(v, s.items, `${pfad}[${i}]`)));
    }
    if (wert !== null && typeof wert === 'object' && !Array.isArray(wert)) {
      for (const p of s.required ?? []) {
        if (!(p in wert)) fehler.push(`${pfad}: Pflichtfeld ${p} fehlt`);
      }
      const props = s.properties ?? {};
      const zusatz = s.additionalProperties;
      for (const [k, v] of Object.entries(wert)) {
        if (k in props) fehler.push(...pruefe(v, props[k], `${pfad}.${k}`));
        else if (zusatz && typeof zusatz === 'object') fehler.push(...pruefe(v, zusatz, `${pfad}.${k}`));
        else if (zusatz === false) fehler.push(`${pfad}: unbekanntes Feld ${k}`);
      }
    }
    return fehler;
  };
  return pruefe(daten, teilschema, '$');
}
